File: mobile/lib/models/weather.dart
/// Weather model (`WeatherOut`).
class Weather {
  const Weather({
    required this.location,
    required this.temperature,
    required this.humidity,
    required this.condition,
    required this.windSpeed,
    required this.updatedAt,
  });

  final String location;
  final double temperature;
  final int humidity;
  final String condition;
  final double windSpeed;
  final String updatedAt;

  factory Weather.fromJson(Object? json) {
    final map = json as Map<String, dynamic>;
    return Weather(
      location: map['location'] as String? ?? '',
      temperature: (map['temperature'] as num?)?.toDouble() ?? 0,
      humidity: (map['humidity'] as num?)?.toInt() ?? 0,
      condition: map['condition'] as String? ?? '',
      windSpeed: (map['wind_speed'] as num?)?.toDouble() ?? 0,
      updatedAt: map['updated_at'] as String? ?? '',
    );
  }
}

File: config/core/models/weather.py
"""
KisanAI OS
Weather Model
Version: 2.0.0
"""

from datetime import datetime

from sqlalchemy import Float, Integer, String, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from config.core.database import Base


def _now():
    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")


class Weather(Base):
    """Weather Model"""

    __tablename__ = "weather"
    __table_args__ = (
        UniqueConstraint("location", name="uq_weather_location"),
    )

    weather_id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    location: Mapped[str] = mapped_column(String(100))
    temperature: Mapped[float] = mapped_column(Float)
    humidity: Mapped[int] = mapped_column(Integer)
    condition: Mapped[str] = mapped_column(String(50))
    wind_speed: Mapped[float] = mapped_column(Float)
    updated_at: Mapped[str] = mapped_column(String(19), default=_now)

    def to_dict(self):
        return {
            "location": self.location,
            "temperature": self.temperature,
            "humidity": self.humidity,
            "condition": self.condition,
            "wind_speed": self.wind_speed,
            "updated_at": self.updated_at,
        }
